### DISCUSSION TITLE
Upload downloaded third-party offline.

### DISCUSSION THREAD

**Discussion by HawtinZeng** [verdaccio/verdaccio#4925] [Ideas]
Hi, verdaccio team:
    Appreciate your hard working, this project has been designed meticulously. Recently I changed a job, which can only work at a offline PC, so I have to set up my own registry with all dependencies, not only my private package. Firstly ,I use the store cache to transfer third party packages, but I cannot search them. Besides, some nested dependencies cannot be cached. So I implement my own 'publish' entry in the web, this button can upload node_modules or a built package or a group of packages(@xxx/ directory in node_modules). the process behind of this button is: 1. generate tar gzip file. 2. generate manifest object. 3. call publish node-api. I haven't implement the integrity code generation, so we need to use `npm install --no-verify` to skip integrity check. Besides, the upload process must be excuted in secure context(in localhost or in https address)
![Recording 2024-10-08 at 21 52 33](https://github.com/user-attachments/assets/6efe3829-691f-4e7a-9f1d-b8b66402d9d9)
The modified code is here, welcome have a try for offline registry setup !

**Comment by HawtinZeng** [comment]
The coresponding docker image is here: https://hub.docker.com/repository/docker/hawtinzeng/verdaccio/general

**Comment by mbtools** [comment]
There's a plugin for the "offline scenario": https://www.npmjs.com/package/verdaccio-offline-storage

The missing download of nested dependencies (only scoped ones), is a known issue
https://github.com/verdaccio/verdaccio/issues/4070 (I'm working on it)

**Comment by mbtools** [comment]
One more thing: By default, npm caches HTTP requests. This means npm will not request a package/tarball from Verdaccio if it is already cached. If you want to make sure everything is downloaded to Verdaccio, use the following:

````sh
npm config set prefer-online true
````

**Comment by huangyiman** [comment]
Hello, I have the same problem as you, very troubled. Have you solved the problem?

**Reply by mbtools** [reply]
check the plugin mentioned above https://github.com/orgs/verdaccio/discussions/4925#discussioncomment-11119362
